### ISSUE TITLE
How can I get the rotation matrix?

### ISSUE THREAD

**Issue by zhangchbin** [nutonomy/nuscenes-devkit#612]
Hi, I am sorry to disturb you. I am in trouble that the pose matrix between the two frames I get is incorrect.
My code is:
```
def get_translation_matrix(translation_vector):
    T = torch.zeros(translation_vector.shape[0], 4, 4)
    t = translation_vector.contiguous().view(-1, 3, 1)
    T[:, 0, 0] = 1
    T[:, 1, 1] = 1
    T[:, 2, 2] = 1
    T[:, 3, 3] = 1
    T[:, :3, 3, None] = t
    return T

def rot_from_qual(a):
    # a: [batchsize, 1, 4]
    w, x, y, z = a[:, 0, 0], a[:, 0, 1], a[:, 0, 2], a[:, 0, 3]
    rot_matrix = torch.zeros((a.shape[0], 3, 3))
    rot_matrix[:, 0, 0] = 1 - 2 * y**2 - 2 * z**2
    rot_matrix[:, 0, 1] = 2 * x * y - 2 * z * w
    rot_matrix[:, 0, 2] = 2 * x * z + 2 * y * w
    rot_matrix[:, 1, 0] = 2 *x * y + 2 * z * w
    rot_matrix[:, 1, 1] = 1 - 2 * x **2 - 2* z**2
    rot_matrix[:, 1, 2] = 2 * y * z - 2 * x * w
    rot_matrix[:, 2, 0] = 2 * x * z - 2 * y * w
    rot_matrix[:, 2, 1] = 2 * y * z + 2 * x * w
    rot_matrix[:, 2, 2] = 1 - 2 * x**2 - 2* y**2
    return rot_matrix

def transformation_from_parameters(axisangle, translation):
    t = translation.clone()
    R_tmp = rot_from_qual(axisangle)
    R = torch.zeros((R_tmp.shape[0], 4, 4))
    R[:, :3, :3] = R_tmp[:, :, :]
    R[:, 3, :3] = 0
    R[:, :3, 3] = t.squeeze(1)[:, :]
    R[:, 3, 3] = 1
    
    T = get_translation_matrix(t)
    M = torch.matmul(T, R)
    return M

if __name__ == '__main__':
    l_axisangle, l_translation, r_axisangle, r_translation = data       # axisangle: [batchsize, 1, 1, 4], translation: [batchsize, 1, 1, 4]
    T_0_n_1 = transformation_from_parameters( l_axisangle[:, 0], l_translation[:, 0])
    T_0_n = transformation_from_parameters(r_axisangle[:, 0], r_translation[:, 0])
    T = torch.bmm(T_0_n, torch.inverse(T_0_n_1))  # get the pose matrix from l -> r
```

**Comment by holger-motional** [comment]
Hi. We use `pyquaternion` for rotations. Please take a look at http://kieranwynn.github.io/pyquaternion/. It should do everything for you.

**Comment by zhangchbin** [comment]
Hi, @holger-motional 
Thanks for your help. How can I get the transform matrix? Am I right?
```
# rotation is quaternion, [q0, q1, q2, q3]
T = Quaternion(rotation).transform_matrix
T[:3, 3] = translation[0, :]
```

**Comment by holger-motional** [comment]
I haven't run it, but it looks right to me. Maybe you create a quaternion with e.g. 90 degree rotation around the z axis and see if the result makes sense.

**Comment by zhangchbin** [comment]
Hi, @holger-motional ,
How can I get point cloud from predicted depth and K ? Can I directly use the code of monodepth2?
The function is here: https://github.com/nianticlabs/monodepth2/blob/master/layers.py#L139-L168

**Comment by holger-motional** [comment]
Hi @zhangchbin. Unfortunately I am not familiar with that codebase.
